scanpy/preprocessing/_scrublet/core.py

```diff
@@ -0,0 +1,466 @@
+from __future__ import annotations
+
+import sys
+from dataclasses import InitVar, dataclass, field
+from typing import TYPE_CHECKING, cast
+
+import numpy as np
+from scipy import sparse
+
+from ... import logging as logg
+from ..._utils import AnyRandom, get_random_state
+from .._utils import sample_comb
+from .neighbors import AnnoyDist, get_knn_graph
+from .sparse_utils import subsample_counts
+
+if TYPE_CHECKING:
+    from numpy.random import RandomState
+    from numpy.typing import NDArray
+
+__all__ = ["Scrublet"]
+
+
+if sys.version_info > (3, 10):
+    kw_only = lambda yes: {"kw_only": yes}  # noqa: E731
+else:
+    kw_only = lambda _: {}  # noqa: E731
+
+
+@dataclass(**kw_only(True))
+class Scrublet:
+    """\
+    Initialize Scrublet object with counts matrix and doublet prediction parameters
+
+    Parameters
+    ----------
+    counts_obs
+        Matrix with shape (n_cells, n_genes) containing raw (unnormalized)
+        UMI-based transcript counts.
+        Converted into a :class:`scipy.sparse.csc_matrix`.
+
+    total_counts_obs
+        Array with shape (n_cells,) of total UMI counts per cell.
+        If `None`, this is calculated as the row sums of `counts_obs`.
+
+    sim_doublet_ratio
+        Number of doublets to simulate relative to the number of observed
+        transcriptomes.
+
+    n_neighbors
+        Number of neighbors used to construct the KNN graph of observed
+        transcriptomes and simulated doublets.
+        If `None`, this is set to round(0.5 * sqrt(n_cells))
+
+    expected_doublet_rate
+        The estimated doublet rate for the experiment.
+
+    stdev_doublet_rate
+        Uncertainty in the expected doublet rate.
+
+    random_state
+        Random state for doublet simulation, approximate
+        nearest neighbor search, and PCA/TruncatedSVD.
+    """
+
+    # init fields
+
+    counts_obs: InitVar[
+        sparse.csr_matrix | sparse.csc_matrix | NDArray[np.integer]
+    ] = field(**kw_only(False))
+    total_counts_obs: InitVar[NDArray[np.integer] | None] = None
+    sim_doublet_ratio: float = 2.0
+    n_neighbors: InitVar[int | None] = None
+    expected_doublet_rate: float = 0.1
+    stdev_doublet_rate: float = 0.02
+    random_state: InitVar[AnyRandom] = 0
+
+    # private fields
+
+    _n_neighbors: int = field(init=False, repr=False)
+    _random_state: RandomState = field(init=False, repr=False)
```

Comment:
But this is used in non-private contexts? https://github.com/scverse/scanpy/pull/2703/files#diff-a42cf1237551ce149c338478446ae890441dc3871397ee52fd5af061fa25ba3eR436-R438

Reply:
This is true of other things as well, but in this case this one used to be public and making/using the private namespace created more diff-bloat actually.